### PR TITLE
Create Involution() module

### PR DESCRIPTION
**Pull request by glenn-jocher**
Involution module PR from https://arxiv.org/abs/2103.06255. Not sure exactly where or how to insert this but getting this PR started to experiment. Github: https://github.com/d-li14/involution

```python
import torch
from models.experimental import Involution

m = Involution(ch=128, k=1, s=1)   # ch_in, kernel, stride
x = torch.zeros(16, 128, 20, 20)  # input
y = m(x)  # forward
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
New Involution layer introduced to YOLOv5 model architecture 🌐

### 📊 Key Changes
- 💡 Implemented a new neural network module called `Involution`, based on a recent research paper.
- 🧠 Added complexity with inner workings like channel grouping, kernel generation, and a unique forward pass mechanism.
- 🔨 Modified the `experimental.py` file to include the Involution class with initialization and forward pass methods.

### 🎯 Purpose & Impact
- 🎛️ Introduces an alternative to convolution, which can adaptively generate weights from the input feature map, potentially enhancing the model's ability to capture intricate spatial patterns.
- 📈 May improve the performance of YOLOv5 by capturing more flexible and diverse features without significantly increasing computational cost.
- 🚀 Provides researchers and practitioners with a new tool for experimenting with state-of-the-art techniques in object detection models.

### CODE REVIEW

**Comment by cszer** [comment]
👍👍👍

**Comment by cszer** [comment]
Hehe seems good
Model Summary: 638 layers, 14631357 parameters, 14631357 gradients, 38.7 GFLOPS

**Comment by cszer** [comment]
Note , that in authors implementation they used involution only when k=3 s=1 , in_channels == out_channels , to use this module change bottelneck to resnet style 

Important - use ks 7 instead of 3 in involution

**Comment by cszer** [comment]
Tested : naive implementation is useless, batch=2 requires 25 gb)) and loss goes to NaN. I have tried cuda but something wrong with compiling, can you try cuda implementation on your hardware?

**Comment by cszer** [comment]
Also i off amp scaler and some other things related to it, always got NaN

**Comment by glenn-jocher** [comment]
@cszer I didn't try to train, but I profiled Involution() against a models.common.Conv() and the default pytorch nn.Conv2d() with some odd results. Even though parameters are much reduced, forward speed is 5x slower and backwards speed about 3x slower.

```python
!git checkout involution

# Profile
from utils.torch_utils import profile 
from models.experimental import Conv, Involution

m1 = Conv(c1=128, c2=128, k=3)
m2 = torch.nn.Conv2d(128, 128, kernel_size=3)
m3 = Involution(ch=128, k=3, s=1)

profile(x=torch.randn(16, 128, 80, 80), ops=[m1, m2, m3], n=100)
```

Output
```
1.8.0+cu101 cuda _CudaDeviceProperties(name='Tesla P100-PCIE-16GB', major=6, minor=0, total_memory=16280MB, multi_processor_count=56)

      Params      GFLOPS    forward (ms)   backward (ms)                   input                  output
      147712       30.25           2.379           5.292       (16, 128, 80, 80)       (16, 128, 80, 80)
      147584           0           2.216            4.53       (16, 128, 80, 80)       (16, 128, 78, 78)
        6536       1.339           11.48           17.18       (16, 128, 80, 80)       (16, 128, 80, 80)
```

**Comment by Dontfall** [comment]
> Tested : naive implementation is useless, batch=2 requires 25 gb)) and loss goes to NaN. I have tried cuda but something wrong with compiling, can you try cuda implementation on your hardware?

Have you conplied it with involution_cuda.py?
/det/mmdet/models/utils/involution_cuda.py

**Comment by QY1994-0919** [comment]
> but I profiled Involution() against a models.common.Conv() and the default pytorch nn.Conv2d() with some odd results

@glenn-jocher 
Hello, I have some problems in the process of integrating evolution into yolov5
1. First, modify the focus part of the backbone in the .Yaml file  
  backbone:
  # [from, number, module, args]
  [[-1, 1, Focus, [64, 3]],  # 0-P1/2
   #[-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 1, Involution, [128, 1, 1]],
2. Secondly, according to your steps experiental.py Create evolution() module in the file,
3. Modify correspondence common.py The modification code of the corresponding replacement part in the file is as follows:
class Focus(nn.Module):
    def __init__(self, c1, c2, k=1, s=1, p=None, g=None, act=True):  # ch_in, ch_out, kernel, stride, padding, groups
        super(Focus, self).__init__()
        self.conv = Involution(Conv(c1 * 4, k, s)
4. But I don't know how to use the following code		
import torch
from models.experimental import Involution
m = Involution(ch=128, k=1, s=1)   # ch_in, kernel, stride
x = torch.zeros(16, 128, 20, 20)  # input
y = m(x)  # forward
I don't know what's wrong with my modification. Can I point it out? thank you

> Involution module PR from https://arxiv.org/abs/2103.06255. Not sure exactly where or how to insert this but getting this PR started to experiment. Github: https://github.com/d-li14/involution
> 
> ```python
> import torch
> from models.experimental import Involution
> 
> m = Involution(ch=128, k=1, s=1)   # ch_in, kernel, stride
> x = torch.zeros(16, 128, 20, 20)  # input
> y = m(x)  # forward
> ```
